### PR TITLE
Proof of concept for MaxDispatches implementation

### PR DESCRIPTION
**Pull request by bentoi**
This PR shows how `MaxDispatches` can be implemented on C#. The implementation for other language mappings would pretty much be the same. Relying on `MaxDispatches` for back pressure also requires to implement `MaxConnections`.

The relationship between `MaxDispatches` and the thread pool max size isn't obvious with this new property. The optimal thread pool maximum size would need to be something like `MaxDispatches * MaxConnections + CPU_Count_Thread_ForIO`.

With IceRPC, we don't have dedicated thread pools so we basically assume that the .NET thread pool has an infinite size.

One possibility could be to make `MaxDispatches/MaxConnections` incompatible with the thread pool properties. You can't set both. If `MaxDispatches/MaxConnections` are set, we configure the thread pool size with a large enough number of threads to allow both IO and dispatches to work (with C# we could even use the .NET thread pool). 

This doesn't solve however the handling of thread starvation for the idle timeout implementation when still using the thread pool configuration...

### CODE REVIEW

**Comment by bernardnormier** [comment]
I opened PR #2603 to just add the MaxDispatches property.

File: csharp/src/Ice/ConnectionI.cs [hunk 1]
```diff
@@ -2516,6 +2532,12 @@ private void sendResponse(OutgoingResponse response, bool isTwoWay, byte compres
                         sendMessage(new OutgoingMessage(response.outputStream, compress > 0, adopt: true));
                     }
 
+                    if (_maxDispatches > 0 && _state != StateHolding && _dispatchCount == _maxDispatches)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:2535]: I am not following here. The _maxDispatches is "active" only state = StateHolding? Should it be only when state = StateActive?

File: csharp/src/Ice/ConnectionI.cs [hunk 2]
```diff
@@ -969,6 +969,12 @@ public override void message(ThreadPoolCurrent current)
                             }
                         }
 
+                        if (_maxDispatches > 0 && _dispatchCount == _maxDispatches)
+                        {
+                            // Only read from the connection if max dispatches isn't reached.
+                            newOp &= ~SocketOperation.Read;
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:975]: This removes Read from newOp, right?

File: csharp/src/Ice/ConnectionI.cs [hunk 3]
```diff
@@ -969,6 +969,12 @@ public override void message(ThreadPoolCurrent current)
                             }
                         }
 
+                        if (_maxDispatches > 0 && _dispatchCount == _maxDispatches)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:972]: Should this be `_dispatchCount >= _maxDispatches`, since _dispatchCount can be incremented by more than 1 when we received a batch request?

File: csharp/src/Ice/ConnectionI.cs [hunk 4]
```diff
@@ -1647,31 +1654,40 @@ private void setState(int state)
...
-                    // Can only switch from holding or not validated to
-                    // active.
+                    // Can only switch to active from holding or not validated.
                     //
                     if (_state != StateHolding && _state != StateNotValidated)
                     {
                         return;
                     }
+
+                    if (_maxDispatches > 0 && _dispatchCount == _maxDispatches)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1664]: Probably should be `>=`

File: csharp/src/Ice/ConnectionI.cs [hunk 5]
```diff
@@ -1647,31 +1654,40 @@ private void setState(int state)
...
-                    // holding.
+                    // Can only switch to holding from active or not validated.
                     //
                     if (_state != StateActive && _state != StateNotValidated)
                     {
                         return;
                     }
-                    if (_state == StateActive)
+
+                    if (_maxDispatches > 0 && _dispatchCount == _maxDispatches)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1684]: `>=`

**Review by bernardnormier — COMMENTED** [review]
How do plan to handle batch requests?

With the current code (and a few fixes), I believe you would accept the batch request and stop reading when the dispatch count is >= max dispatches. So with MaxDispatches = 10, I could dispatch a batch request with 20 elements. Is this a desirable/expected behavior?

**Review by bernardnormier — COMMENTED** [review]
> This doesn't solve however the handling of thread starvation for the idle timeout implementation when still using the thread pool configuration...

In my view, once we implement MaxDispatches, we should not worry about the idle check aborting the connection when the thread pool is starved. If you want to apply back pressure, use MaxDispatches, not thread pool starvation.

**Review by bernardnormier — COMMENTED** [review]
> The relationship between MaxDispatches and the thread pool max size isn't obvious with this new property.

I don't think there is any relationship between them.

Say you use AMD, and you have a single thread in your thread pool. You could still easily write a test with 100 outstanding dispatches.

File: csharp/src/Ice/ConnectionI.cs (hunk 1 above)
Comment (bentoi) [reply csharp/src/Ice/ConnectionI.cs:2535]: Did you expect `_state == StateActive` instead? It's pretty much the same.

File: csharp/src/Ice/ConnectionI.cs (hunk 2 above)
Comment (bentoi) [reply csharp/src/Ice/ConnectionI.cs:975]: Yes, I'll move this code to `parseMessage` instead, it should be a little clearer.

**Comment by bentoi** [comment]
> I could dispatch a batch request with 20 elements. Is this a desirable/expected behavior?

It's for sure the simplest to implement and I don't think it's worth implementing something complicated for a corner case scenario like this one.

**Comment by bentoi** [comment]
> Say you use AMD, and you have a single thread in your thread pool. You could still easily write a test with 100 outstanding dispatches.

True but the goal is to ensure that the idle timeout works and for the idle timeout to work you need and thread pool large enough to accommodate with the case where you only have sync dispatches. I wouldn't keep the default single threaded thread pool if `MaxDispatches` is configured. I'd go for a large thread pool max size instead.

**Comment by bernardnormier** [comment]
> the goal is to ensure that the idle timeout works and for the idle timeout to work you need and thread pool large enough to accommodate with the case where you only have sync dispatches. I wouldn't keep the default single threaded thread pool if `MaxDispatches` is configured. I'd go for a large thread pool max size instead.

That's not how I see it.

In a healthy application, you can occasionally hit the thread pool limit for a short amount of time (like a few seconds), and this won't trigger an idle check abort since the default (and typical) idle timeout is 60 seconds.

As of Ice 3.8, an application should not use thread-pool exhaustion to apply back pressure (because the correct tool for that is MaxDispatches). And as a result, the idle check logic doesn't need to account for a legitimate use of thread-pool exhaustion since there isn't one anymore.

Say you upgrade from 3.7 to 3.8 and notice your upgraded application (that doesn't worry about flow control/back pressure) sees idle-check aborts. It's probably a sign you have long & unplanned thread pool exhaustions and you want to increase your thread pool size.

**Comment by bernardnormier** [comment]
> > I could dispatch a batch request with 20 elements. Is this a desirable/expected behavior?
> 
> It's for sure the simplest to implement and I don't think it's worth implementing something complicated for a corner case scenario like this one.

I believe we need to handle this limit properly. 

Say you set MaxDispatches=10 as a "security" measure, to make sure your client doesn't bombard you with concurrent requests. But then a client can easily bypass this limit by crafting a batch with 100 oneway requests? This doesn't make sense.

File: csharp/src/Ice/Internal/Instance.cs [hunk 6]
```diff
@@ -1436,7 +1435,11 @@ internal ConnectionOptions serverConnectionOptions(string adapterName)
 
                 inactivityTimeout: TimeSpan.FromSeconds(properties.getPropertyAsIntWithDefault(
                     $"{adapterName}.Connection.InactivityTimeout",
-                    (int)clientConnectionOptions.inactivityTimeout.TotalSeconds)));
+                    (int)clientConnectionOptions.inactivityTimeout.TotalSeconds)),
+
+                maxDispatches: properties.getPropertyAsIntWithDefault(
+                    $"{adapterName}.Connection.MaxDispatches",
+                    (int)clientConnectionOptions.maxDispatches));
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/Instance.cs:1442]: no need to cast an int to an int.

File: csharp/src/Ice/ConnectionI.cs [hunk 7]
```diff
@@ -1647,31 +1648,40 @@ private void setState(int state)
...
-                    if (_state == StateActive)
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
                     {
-                        _threadPool.unregister(this, SocketOperation.Read);
+                        // Reads are already disabled if the max dispatch count is reached.
+                        return;
                     }
+
+                    _threadPool.unregister(this, SocketOperation.Read);
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1684]: There is an extra change compared to the old code:
 - old code, state = StateNotValidated, we don't unregister and just 'break'
 - new code, state = StateNotValidated treated like StateActive

 Is this intentional?

File: csharp/src/Ice/ConnectionI.cs [hunk 8]
```diff
@@ -1647,31 +1648,40 @@ private void setState(int state)
...
                     _threadPool.register(this, SocketOperation.Read);
                     break;
                 }
 
                 case StateHolding:
                 {
                     //
-                    // Can only switch from active or not validated to
-                    // holding.
+                    // Can only switch to holding from active or not validated.
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1671]: This comment is confusing. If this is an impossible state transition, we should assert, not "if ... return".

File: csharp/src/Ice/ConnectionI.cs [hunk 9]
```diff
@@ -2515,6 +2533,13 @@ private void sendResponse(OutgoingResponse response, bool isTwoWay, byte compres
                         sendMessage(new OutgoingMessage(response.outputStream, compress > 0, adopt: true));
                     }
 
+                    if (_state == StateActive && _maxDispatches > 0 && _dispatchCount == _maxDispatches)
+                    {
+                        // Resume reads if the connection is active and the dispatch count is (about to be) bellow the
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:2538]: below

File: csharp/src/Ice/ConnectionI.cs [hunk 10]
```diff
@@ -2515,6 +2533,13 @@ private void sendResponse(OutgoingResponse response, bool isTwoWay, byte compres
...
+                        // max dispatch count.
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:2539]: Do you want to rename the field _maxDispatchCount?

File: csharp/src/Ice/Internal/ConnectionFactory.cs [hunk 11]
```diff
@@ -1419,6 +1419,20 @@ public override void message(ThreadPoolCurrent current)
                 }
 
                 _connections.Add(connection);
+                if (_maxConnections > 0 && _connections.Count == _maxConnections)
+                {
+                    if (_instance.traceLevels().network >= 1)
+                    {
+                        StringBuilder s = new StringBuilder("holding ");
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/ConnectionFactory.cs:1426]: In modern C#, you should use an interpolated string:
```csharp
string message = $"holding {_endpoint.protocol()} connections at {_acceptor}...";
```

File: csharp/src/Ice/Internal/ConnectionFactory.cs [hunk 12]
```diff
@@ -1504,6 +1518,9 @@ public void connectionStartFailed(Ice.ConnectionI connection, Ice.LocalException
         _state = StateHolding;
         _acceptorStarted = false;
 
+        // TODO: add MaxConnections property.
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/ConnectionFactory.cs:1521]: I am confused, you already added this property to ObjectAdapter.

File: csharp/src/Ice/Internal/ConnectionFactory.cs [hunk 13]
```diff
@@ -1595,6 +1612,13 @@ private void setState(int state)
                 {
                     return;
                 }
+
+                if (_maxConnections > 0 && _connections.Count == _maxConnections)
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/ConnectionFactory.cs:1616]: Say the application calls `adapter.activate()` on an OA in the holding state.

If at this point `_connections.Count == _maxConnections`, the state of the adapter remains Holding. Should we throw an exception to notify the application `adapter.activate()` was not successful?

Moreover:
 - holding = don't accept new incoming connections AND don't read from existing (incoming) connections.
 - max connection reached = don't accept new incoming connections but keep reading from existing incoming connections

So I think we should actually transition from Holding to Active here, and resume reading from the existing connections. Just don't accept new connections as long as `_connections.Count == _maxConnections`.

**Comment by bentoi** [comment]
> a client can easily bypass this limit by crafting a batch with 100 oneway requests? This doesn't make sense.

On the wire, 100 oneways requests isn't much heavier than a batch of 100 oneway requests. Batch oneway requests are also serialized for dispatching. 

The main issue is for asynchronous oneway dispatch since the lifetime of a dispatch is bounded by the completion of the AMD callback. Or do you see any other scenarios where it's a problem?

File: csharp/src/Ice/ConnectionI.cs (hunk 10 above)
Comment (bentoi) [reply csharp/src/Ice/ConnectionI.cs:2539]: I don't have a preference. I'm fine with `_maxDispatchCount` if you prefer it.

File: csharp/src/Ice/ConnectionI.cs (hunk 7 above)
Comment (bentoi) [reply csharp/src/Ice/ConnectionI.cs:1684]: It doesn't make a difference. If `state == StateNotValidated`, unregistering reads is fine, it will be a no-op. The previous code assumes reads were already unregistered -- it wasn't bogus because it's indeed the case.

I find it clearer to have the same code as for the `StateActive` case.

**Comment by bernardnormier** [comment]
> On the wire, 100 oneways requests isn't much heavier than a batch of 100 oneway requests.

The whole point of batching is to make this 100-batch message smaller/more efficient than 100 individual oneway requests.

> Batch oneway requests are also serialized for dispatching.

That's actually not necessarily true. Take our latest Swift code, where we launch a Task for each incoming request. Unless we put some limit, all the requests in a batch can be dispatched concurrently.
 
> The main issue is for asynchronous oneway dispatch since the lifetime of a dispatch is bounded by the completion of the AMD callback. Or do you see any other scenarios where it's a problem?

Yes, the issue is with AMD dispatches. In Swift (on main) this is the only kind of dispatch.

I think we can add some batch-specific code in "dispatchAll" to make sure we don't dispatch more than _maxDispatches concurrently when _dispatchCount > _maxDispatches.

File: csharp/src/Ice/ConnectionI.cs (hunk 10 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:2539]: I prefer _maxDispatches like the property name. We should use the same terminology in the comments.

File: csharp/src/Ice/ConnectionI.cs (hunk 7 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:1684]: Ok

File: csharp/src/Ice/Internal/ConnectionFactory.cs (hunk 13 above)
Comment (bentoi) [reply csharp/src/Ice/Internal/ConnectionFactory.cs:1616]: I agree.

File: csharp/src/Ice/Internal/ConnectionFactory.cs (hunk 12 above)
Comment (bentoi) [reply csharp/src/Ice/Internal/ConnectionFactory.cs:1521]: I don't think I did but I'll add it.

**Comment by bentoi** [comment]
> The whole point of batching is to make this 100-batch message smaller/more efficient than 100 individual oneway requests.

I don't think size on the wire was the only reason for supporting batch messages. A "send" socket call was considered an expensive call at the time Ice was written. Batching oneway messages allowed to make a single "send" call instead of many. See https://doc.zeroc.com/ice/3.7/client-side-features/batched-invocations

> That's actually not necessarily true. Take our latest Swift code, where we launch a Task for each incoming request. Unless we put some limit, all the requests in a batch can be dispatched concurrently.

Serializing the dispatch of oneway from a batch is a documented guarantee (on the page mentioned above): _"On the server side, batched messages are dispatched by a single thread, in the order in which they were written into the batch. This means that messages from a single batch cannot appear to be reordered in the server."_

What was the reason for not providing this guarantee in Swift?

> I think we can add some batch-specific code in "dispatchAll" to make sure we don't dispatch more than _maxDispatches concurrently when _dispatchCount > _maxDispatches.

I'll look into it.

**Comment by bentoi** [comment]
> > I think we can add some batch-specific code in "dispatchAll" to make sure we don't dispatch more than _maxDispatches concurrently when _dispatchCount > _maxDispatches.
> 
> I'll look into it.

I don't see how we can implement this and keep the guarantee that oneway requests from a batch are dispatched with the same thread pool thread... Is your idea to hold on the thread with a sync semaphore wait?

To me, the ordering guarantee is more important than using the same thread so one option could be to await the `dispatchAsync(request)` call. 

I don't think the comment about the continuation running on the .NET thread pool from `dispatchAll` is actually correct. The continuation should in theory run a thread pool thread because the thread pool sets a synchronization context, see https://github.com/zeroc-ice/ice/blob/1ae9e22a64a2f8c80bdf5ad280f71064a4000c64/csharp/src/Ice/Internal/ThreadPool.cs#L796

Changing this will change the AMD dispatch of oneways from a batch however... since the next oneway won't be dispatch before the AMD callback is completed.

**Comment by bernardnormier** [comment]
> Is your idea to hold on the thread with a sync semaphore wait?

No, I don't think we should block the dispatch thread to implement MaxDispatches with batches.

> To me, the ordering guarantee is more important than using the same thread so one option could be to await the dispatchAsync(request) call.

I think you're reading too much into this guarantee, and how important it is to keep it. We documented an implementation detail and now we're stuck with it?

How would you provide this guarantee in Swift?

As a reminder, in Swift, each dispatch corresponds to the creation of a Task executed by the Swift global concurrent executor. And there is no limit / serialization other than the future MaxDispatches. So N oneway requests in a batch are dispatched exactly like N oneway requests in separate requests.

**Comment by bernardnormier** [comment]
> Changing this will change the AMD dispatch of oneways from a batch however... since the next oneway won't be dispatch before the AMD callback is completed.

I don't think we should change this behavior. And yes, we should fix the comment (in ConnectionI.cs) since it's inaccurate.

**Comment by bentoi** [comment]
> I think you're reading too much into this guarantee, and how important it is to keep it. We documented an implementation detail and now we're stuck with it?

To me we document a guarantee that we put forward in the past :slightly_smiling_face:. Batching is useful to ensure multiple client invocations are executed in same order on the server. Without it, you need a single threaded server or serialized twoway invocations.

Do we have clients relying on this? Probably not... I doubt many clients actually use batching. Glacier2/IceStorm do use it however (to reduce the number of oneway messages to forwarded by batching them).

> How would you provide this guarantee in Swift?

That's an implementation detail :slightly_smiling_face:.

> I don't think we should change this behavior. And yes, we should fix the comment (in ConnectionI.cs) since it's inaccurate.

Actually, I don't think my previous comment about the synchronization context is correct since we call `ConfigureAwait(false)` on `dispatcher.dispatchAsync()`...  We never used await/async in the Ice implementation before and we didn't have to worry about this. The cohabitation of the Ice thread pool and the .NET thread pool is really messy...

>> Changing this will change the AMD dispatch of oneways from a batch however... since the next oneway won't be dispatch before the AMD callback is completed.
> I don't think we should change this behavior.

I still don't see how to limit the number of dispatch with batch requests. The servant needs to unmarshal the input parameters of a oneway message before we can dispatch the next one. 

Waiting for a batch request to be totally dispatched (i.e.: once the AMD callback is called) sounds like the simplest (only?) solution to me. Wouldn't it also provide the ordering guarantee for Swift?

**Comment by bernardnormier** [comment]
After discussing this issue with Joe, we agree with you:
 - let's ignore this "batch issue": when we receive a batch request, we can go over MaxDispatches
 - we'll fix C++ & Swift to add special batch handling

**Comment by bernardnormier** [comment]
Replaced by #2672 and #2676.
